File: Algorithms/0048_Rotate_Image/Java/RotateImageSolution1.java

// Space: O(1)
// Time: O(n)

class Solution {
    public void rotate(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        if (m == 1 && n == 1) return;

        for (int x = 0; x < n - 1; x++) {
            int offset = 1;
            for (int y = m - 2 - x; y >= 0; y--) {
                int temp = matrix[y][x];
                matrix[y][x] = matrix[y + offset][x + offset];
                matrix[y + offset][x + offset] = temp;
                offset++;
            }
        }

        for (int y = 0; y < m / 2; y++) {
            for (int x = 0; x < n; x++) {
                int temp = matrix[y][x];
                matrix[y][x] = matrix[m - 1 - y][x];
                matrix[m - 1 - y][x] = temp;
            }
        }
    }
}

File: Algorithms/0048_Rotate_Image/Python/Rotate_Image_Solution_1.py

# Space: O(1)
# Time: O(n)

class Solution:
    def rotate(self, matrix):
        """
        Do not return anything, modify matrix in-place instead.
        """
        m = len(matrix)
        n = len(matrix[0])
        if m == 1 and n == 1: return

        for x in range(n - 1):
            offset = 1
            for y in range(m - 2 - x, -1, -1):
                matrix[y][x], matrix[y + offset][x + offset] = matrix[y + offset][x + offset], matrix[y][x]
                offset += 1

        for y in range(m // 2):
            for x in range(n):
                matrix[y][x], matrix[m - y - 1][x] = matrix[m - y - 1][x], matrix[y][x]




